### ISSUE TITLE
pass generated code through formatter in Paste Java as Ceylon

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-ide-eclipse#1532]
Paste Java as Ceylon produces code with the indentation all screwed up.

**Comment by lucaswerkmeister** [comment]
IIRC there’s code that’s *supposed* to do this, but it doesn’t work for some reason. Haven’t had the time to look at it yet.

**Comment by lucaswerkmeister** [comment]
https://github.com/ceylon/ceylon-ide-eclipse/blob/0ea005c8c97fe3eaf14b7a6bdf8c81c778ed714a/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/code/editor/PasteAsCeylonHandler.java#L98-L99

**Comment by gavinking** [comment]
@lucaswerkmeister well it seems obvious why it's not working, right? It's using the tokens from the previously-parsed source, instead of parsing the pasted text.

**Comment by gavinking** [comment]
@lucaswerkmeister don't you have a function to just format a `String` containing source code?

**Comment by lucaswerkmeister** [comment]
Ah, that makes sense… the `DocumentChange` doesn’t cause a re-parse, right?

**Comment by lucaswerkmeister** [comment]
> don't you have a function to just format a String containing source code?

Not directly, no. But it’s not hard to rig up a new `CeylonLexer` with an `AntlrStringStream`.

**Comment by lucaswerkmeister** [comment]
This is the version I use in `ceylon.ast`:
```ceylon
CeylonParser(CommonTokenStream(CeylonLexer(ANTLRStringStream(code + " "))))
```
Calling `.compilationUnit()` on that should give us an AST that the formatter can take. (Probably extract the token source and reset it and give a `BufferedTokenStream` of it to the formatter as well.)

**Comment by gavinking** [comment]
Done.

**Comment by bjansen** [comment]
I noticed that when I convert a single method and paste it in a class/object, the code is formatted but the indentation is not correct, i.e. it generates something like this:

```
class A() {
void function() {
    print(...);
}
}
```

It's like the formatter is not aware we are in a class?

**Comment by lucaswerkmeister** [comment]
@bjansen yeah, [currently it always has the initial indentation `0`](https://github.com/ceylon/ceylon-ide-eclipse/commit/e73e13123bdc9ce9c004b7c89cd74666bc34d480#diff-9531d6b7569df89b1e1737c3e1738c35R144). I’ll try to see if I can get it to use the current indentation instead…

**Comment by bjansen** [comment]
Cool, thanks!

**Comment by lucaswerkmeister** [comment]
@bjansen should be better now. (I also found a bug in the regular `FormatAction`, which didn’t count tabs correctly when determining the indentation level – fixed in c2a36d0e9f124665f427590fe663f083bb8bf139.)
